src/EditorFeatures/Core/GoToBase/AbstractGoToBaseService.cs

```diff
@@ -12,34 +13,60 @@ internal abstract partial class AbstractGoToBaseService : IGoToBaseService
         public async Task FindBasesAsync(Document document, int position, IFindUsagesContext context)
         {
             var cancellationToken = context.CancellationToken;
-            var tuple = await FindBaseHelpers.FindBasesAsync(document, position, cancellationToken).ConfigureAwait(false);
-            if (tuple == null)
+            var symbolAndProject = await FindUsagesHelpers.GetRelevantSymbolAndProjectAtPositionAsync(
+                document, position, cancellationToken).ConfigureAwait(false);
+
+            if (symbolAndProject == default)
             {
                 await context.ReportMessageAsync(
                     EditorFeaturesResources.Cannot_navigate_to_the_symbol_under_the_caret).ConfigureAwait(false);
                 return;
             }
 
-            var (symbol, implementations, message) = tuple.Value;
-
-            if (message != null)
-            {
-                await context.ReportMessageAsync(message).ConfigureAwait(false);
-                return;
-            }
+            var symbol = symbolAndProject.Value.symbol;
+            var bases = FindBaseHelpers.FindBases(
+                symbol, symbolAndProject.Value.project, cancellationToken);
 
             await context.SetSearchTitleAsync(
                 string.Format(EditorFeaturesResources._0_bases,
                 FindUsagesHelpers.GetDisplayName(symbol))).ConfigureAwait(false);
 
-            var solution = document.Project.Solution;
+            var project = document.Project;
+            var solution = project.Solution;
+            var projectId = project.Id;
+
+            var found = false;
+
+            // For each potential base, try to find its definition in sources.
+            // If found, add its' definitionItem to the context.
+            // If not found but the symbol is from metadata, create its' definition item from metadata and add to the context.
+            foreach (var baseSymbol in bases)
+            {
+                var sourceDefinition = await SymbolFinder.FindSourceDefinitionAsync(
+                   SymbolAndProjectId.Create(baseSymbol, projectId), solution, cancellationToken).ConfigureAwait(false);
+                if (sourceDefinition.Symbol != null &&
+                    sourceDefinition.Symbol.Locations.Any(l => l.IsInSource))
```

Comment:
FindSourceDefinitionAsync should only be returning symbols from source in the first place...